buildkite/scripts/lagrange-cache-manager.sh

```diff
@@ -0,0 +1,99 @@
+#!/bin/bash
+set -eo pipefail
+
+# Function to create cache directory with subdirectories for pallas and vesta
+create_cache_dir() {
+  local cache_dir=$1
+  mkdir -p "$cache_dir/pallas" "$cache_dir/vesta"
+}
+
+# Function to restore cache, preserving the directory structure
+restore_cache() {
+  local bucket_name=$1
+  local cache_dir=$2
+
+  echo "Restoring cache from bucket: gs://$bucket_name"
+  
+  if gsutil cp -r "gs://$bucket_name/pallas/*" "$cache_dir/pallas/" && \
+     gsutil cp -r "gs://$bucket_name/vesta/*" "$cache_dir/vesta/"; then
+    echo "Cache restored from GCS."
+  else
+    echo "No cache found. Starting with a fresh cache."
+  fi
+}
+
+# Function to list files in GCS bucket with error handling for empty directories
+list_gcs_files() {
+  local bucket_name=$1
+  local prefix=$2
+  gsutil ls -r "gs://$bucket_name/$prefix/**" 2>/dev/null | sed 's#gs://'$bucket_name'/##' || true
+}
+
+# Function to upload only new or changed files, preserving the directory structure
+upload_cache_if_changed() {
+  local bucket_name=$1
+  local cache_dir=$2
+
+  # List files currently in the bucket, restricted to pallas and vesta directories
+  bucket_files=$(list_gcs_files "$bucket_name" "pallas")
+  bucket_files+=$'\n'
+  bucket_files+=$(list_gcs_files "$bucket_name" "vesta")
+
+  # Create a list of files currently in the cache directory, restricted to pallas and vesta directories
+  local_files=$(find "$cache_dir/pallas" "$cache_dir/vesta" -type f | sed 's#^'$cache_dir'/##')
+
+  echo "Local cache files: $local_files"
+
+  # Find new or changed files by comparing the lists
```

Comment:
I suggest just doing `gsutil cp` with an `-n` flag.

This way `gsutil` will copy all files, but skip copying files that already exist on the bucket.